### ISSUE TITLE
need eyeballs on OS X Gunrock failure on prerelease ctest

### ISSUE THREAD

**Issue by jowens** [gunrock/gunrock#183]
@yzhwang or others, just want a pair of eyeballs on this to tell me if you've seen it before. If not, I'll dig in. This is OS X with CUDA 8 prerelease. It built fine.

(Also I note `NOT COMPILED WITH COMPATIBLE PTX VERSION FOR DEVICE 0`, which has never occurred before. Problem with cmake? @neoblizz ?)

```
test 1
      Start  1: TEST_BFS

1: Test command: /Users/jowens/Documents/working/gunrock-build/bin/bfs "market" "/Users/jowens/Documents/working/gunrock/simple_example/bips98_606.mtx" "--undirected" "--src=0"
1: Test timeout computed to be: 9.99988e+06
1: Loading Matrix-market coordinate-formatted graph ...
1:   Reading directly from stored binary CSR arrays ...
1: Done reading (0s).
1:
1: Degree Histogram (7135 vertices, 30380 edges):
1:     Degree   0: 541 (7.58%)
1:     Degree 2^0: 626 (8.77%)
1:     Degree 2^1: 2181 (30.57%)
1:     Degree 2^2: 3085 (43.24%)
1:     Degree 2^3: 621 (8.70%)
1:     Degree 2^4: 80 (1.12%)
1:     Degree 2^5: 0 (0.00%)
1:     Degree 2^6: 1 (0.01%)
1:
1: [/Users/jowens/Documents/working/gunrock/gunrock/app/bfs/bfs_enactor.cuh, 1865 @ gpu 0] BFSEnactor cudaBindTexture labels_tex_ref failed (CUDA error 18: invalid texture reference)
1: [/Users/jowens/Documents/working/gunrock/tests/bfs/test_bfs.cu, 388 @ gpu 0] BFS Enactor Init failed (CUDA error 18: invalid texture reference)
1: Source vertex: 0
1: Using 1 GPU: [ 0 ].
1: NOT COMPILED WITH COMPATIBLE PTX VERSION FOR DEVICE 0
1:
1:  [BFS] finished.
1:  avg. elapsed: 0.0000 ms
1:  iterations: 0
1:  load time: 165.9155 ms
1:  preprocess time: 0.0000 ms
1:  postprocess time: 0.0000 ms
1:  total time: 169.1387 ms
1:
 1/11 Test  #1: TEST_BFS .........................***Failed  Required regular expression not found.Regex=[CORRECT
]  0.18 sec
```

**Comment by jowens** [comment]
It fails on every test; here's a different one.

```
3: Test command: /Users/jowens/Documents/working/gunrock-build/bin/cc "market" "/Users/jowens/Documents/working/gunrock/simple_example/bips98_606.mtx"
3: Test timeout computed to be: 9.99988e+06
3: Loading Matrix-market coordinate-formatted graph ...
3:   Reading directly from stored binary CSR arrays ...
3: Done reading (0s).
3:
3: Degree Histogram (7135 vertices, 30380 edges):
3:     Degree   0: 541 (7.58%)
3:     Degree 2^0: 626 (8.77%)
3:     Degree 2^1: 2181 (30.57%)
3:     Degree 2^2: 3085 (43.24%)
3:     Degree 2^3: 621 (8.70%)
3:     Degree 2^4: 80 (1.12%)
3:     Degree 2^5: 0 (0.00%)
3:     Degree 2^6: 1 (0.01%)
3:
3: [/Users/jowens/Documents/working/gunrock/gunrock/util/array_utils.cuh, 500 @ gpu 0] vertex_associate_outs movment source is not valid (CUDA error 8: invalid device function)
3: [/Users/jowens/Documents/working/gunrock/tests/cc/test_cc.cu, 363 @ gpu 0] CC Problem Initialization Failed (CUDA error 8: invalid device function)
3: Using 1 GPU: [ 0 ].
3: NOT COMPILED WITH COMPATIBLE PTX VERSION FOR DEVICE 0
3: graph #nodes : 7135 -> 6594
3:
3:  [CC] finished.
3:  avg. elapsed: 0.0000 ms
3:  iterations: 0
3:  load time: 169.5655 ms
3:  preprocess time: 0.0000 ms
3:  postprocess time: 0.0000 ms
3:  total time: 171.6510 ms
3:
 3/11 Test  #3: TEST_CC ..........................***Failed  Required regular expression not found.Regex=[CORRECT
]  0.18 sec
```

**Comment by neoblizz** [comment]
@jowens is this output with verbose? Can you please post output of `make check`? 

Also what GPU does you mac runs on; it will be good to know the driver version as well to figure out the compatibility with CUDA 8.0.

**Comment by jowens** [comment]
Yes, that was verbose. Here is `make check`:

```
$ make check
-- Boost version: 1.59.0
-- Boost version: 1.59.0
-- Found the following Boost libraries:
--   system
--   filesystem
--   timer
--   chrono
-- Found OpenMP (-libomp)
-- Found Metis
-- Project Added: bc
-- Project Added: bfs
-- Project Added: cc
-- Project Added: pr
-- Project Added: sssp
-- Project Added: hits
-- Project Added: salsa
-- Project Added: wtf
-- Project Added: topk
-- Configuring done
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   gunrock

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /Users/jowens/Documents/working/gunrock-build
Scanning dependencies of target check
Test project /Users/jowens/Documents/working/gunrock-build
      Start  1: TEST_BFS
 1/11 Test  #1: TEST_BFS .........................***Failed  Required regular expression not found.Regex=[CORRECT
]  0.26 sec
Loading Matrix-market coordinate-formatted graph ...
  Reading directly from stored binary CSR arrays ...
Done reading (0s).

Degree Histogram (7135 vertices, 30380 edges):
    Degree   0: 541 (7.58%)
    Degree 2^0: 626 (8.77%)
    Degree 2^1: 2181 (30.57%)
    Degree 2^2: 3085 (43.24%)
    Degree 2^3: 621 (8.70%)
    Degree 2^4: 80 (1.12%)
    Degree 2^5: 0 (0.00%)
    Degree 2^6: 1 (0.01%)

[/Users/jowens/Documents/working/gunrock/gunrock/app/bfs/bfs_enactor.cuh, 1865 @ gpu 0] BFSEnactor cudaBindTexture labels_tex_ref failed (CUDA error 18: invalid texture reference)
[/Users/jowens/Documents/working/gunrock/tests/bfs/test_bfs.cu, 388 @ gpu 0] BFS Enactor Init failed (CUDA error 18: invalid texture reference)
Source vertex: 0
Using 1 GPU: [ 0 ].
NOT COMPILED WITH COMPATIBLE PTX VERSION FOR DEVICE 0

 [BFS] finished.
 avg. elapsed: 0.0000 ms
 iterations: 0
 load time: 223.9190 ms
 preprocess time: 0.0000 ms
 postprocess time: 0.0000 ms
 total time: 234.7229 ms


      Start  2: TEST_BC
 2/11 Test  #2: TEST_BC ..........................***Failed  Required regular expression not found.Regex=[CORRECT
]  0.20 sec
Loading Matrix-market coordinate-formatted graph ...
  Reading directly from stored binary CSR arrays ...
Done reading (0s).

Degree Histogram (7135 vertices, 30380 edges):
    Degree   0: 541 (7.58%)
    Degree 2^0: 626 (8.77%)
    Degree 2^1: 2181 (30.57%)
    Degree 2^2: 3085 (43.24%)
    Degree 2^3: 621 (8.70%)
    Degree 2^4: 80 (1.12%)
    Degree 2^5: 0 (0.00%)
    Degree 2^6: 1 (0.01%)

[/Users/jowens/Documents/working/gunrock/gunrock/app/bc/bc_enactor.cuh, 1905 @ gpu 0] BFSEnactor cudaBindTexture row_offsets_ref failed (CUDA error 18: invalid texture reference)
[/Users/jowens/Documents/working/gunrock/tests/bc/test_bc.cu, 502 @ gpu 0] BC Enactor init failed (CUDA error 18: invalid texture reference)
Source vertex: 0
Using 1 GPU: [ 0 ].
NOT COMPILED WITH COMPATIBLE PTX VERSION FOR DEVICE 0

 [BC] finished.
 avg. elapsed: 0.0000 ms
 iterations: 0
 load time: 168.9948 ms
 preprocess time: 0.0000 ms
 postprocess time: 0.0000 ms
 total time: 177.2486 ms


      Start  3: TEST_CC
 3/11 Test  #3: TEST_CC ..........................***Failed  Required regular expression not found.Regex=[CORRECT
]  0.22 sec
Loading Matrix-market coordinate-formatted graph ...
  Reading directly from stored binary CSR arrays ...
Done reading (0s).

Degree Histogram (7135 vertices, 30380 edges):
    Degree   0: 541 (7.58%)
    Degree 2^0: 626 (8.77%)
    Degree 2^1: 2181 (30.57%)
    Degree 2^2: 3085 (43.24%)
    Degree 2^3: 621 (8.70%)
    Degree 2^4: 80 (1.12%)
    Degree 2^5: 0 (0.00%)
    Degree 2^6: 1 (0.01%)

[/Users/jowens/Documents/working/gunrock/gunrock/util/array_utils.cuh, 500 @ gpu 0] vertex_associate_outs movment source is not valid (CUDA error 8: invalid device function)
[/Users/jowens/Documents/working/gunrock/tests/cc/test_cc.cu, 363 @ gpu 0] CC Problem Initialization Failed (CUDA error 8: invalid device function)
Using 1 GPU: [ 0 ].
NOT COMPILED WITH COMPATIBLE PTX VERSION FOR DEVICE 0
graph #nodes : 7135 -> 6594

 [CC] finished.
 avg. elapsed: 0.0000 ms
 iterations: 0
 load time: 192.4558 ms
 preprocess time: 0.0000 ms
 postprocess time: 0.0000 ms
 total time: 197.7964 ms


      Start  4: TEST_SSSP
 4/11 Test  #4: TEST_SSSP ........................***Failed  Required regular expression not found.Regex=[CORRECT
]  0.22 sec
Loading Matrix-market coordinate-formatted graph ...
  Reading directly from stored binary CSR arrays ...
Done reading (0s).

Degree Histogram (39 vertices, 340 edges):
    Degree   0: 0 (0.00%)
    Degree 2^0: 0 (0.00%)
    Degree 2^1: 1 (2.56%)
    Degree 2^2: 22 (56.41%)
    Degree 2^3: 13 (33.33%)
    Degree 2^4: 2 (5.13%)
    Degree 2^5: 1 (2.56%)

[/Users/jowens/Documents/working/gunrock/gunrock/app/sssp/sssp_enactor.cuh, 972 @ gpu 0] BFSEnactor cudaBindTexture row_offsets_ref failed (CUDA error 18: invalid texture reference)
[/Users/jowens/Documents/working/gunrock/tests/sssp/test_sssp.cu, 385 @ gpu 0] SSSP Enactor Init failed (CUDA error 18: invalid texture reference)
Source vertex: 0
Using 1 GPU: [ 0 ].
NOT COMPILED WITH COMPATIBLE PTX VERSION FOR DEVICE 0

 [SSSP] finished.
 avg. elapsed: 0.0000 ms
 iterations: 0
 load time: 177.8257 ms
 preprocess time: 0.0000 ms
 postprocess time: 0.0000 ms
 total time: 188.0836 ms


      Start  5: TEST_PAGERANK
 5/11 Test  #5: TEST_PAGERANK ....................***Failed  Required regular expression not found.Regex=[CORRECT
]  0.21 sec
Loading Matrix-market coordinate-formatted graph ...
  Reading directly from stored binary CSR arrays ...
Done reading (0s).

Degree Histogram (7135 vertices, 30380 edges):
    Degree   0: 541 (7.58%)
    Degree 2^0: 626 (8.77%)
    Degree 2^1: 2181 (30.57%)
    Degree 2^2: 3085 (43.24%)
    Degree 2^3: 621 (8.70%)
    Degree 2^4: 80 (1.12%)
    Degree 2^5: 0 (0.00%)
    Degree 2^6: 1 (0.01%)

[/Users/jowens/Documents/working/gunrock/gunrock/app/pr/pr_enactor.cuh, 1268 @ gpu 0] BFSEnactor cudaBindTexture row_offsets_ref failed (CUDA error 18: invalid texture reference)
[/Users/jowens/Documents/working/gunrock/tests/pr/test_pr.cu, 643 @ gpu 0] PR Enactor Init failed (CUDA error 18: invalid texture reference)
Using 1 GPU: [ 0 ].
NOT COMPILED WITH COMPATIBLE PTX VERSION FOR DEVICE 0

 [PageRank] finished.
 avg. elapsed: 0.0000 ms
 iterations: 0
 load time: 169.3113 ms
 preprocess time: 0.0000 ms
 postprocess time: 0.0000 ms
 total time: 180.7867 ms


      Start  6: TEST_TOPK
 6/11 Test  #6: TEST_TOPK ........................   Passed    0.20 sec
      Start  7: SHARED_LIB_TEST_BFS
```

After this it hangs.

**Comment by jowens** [comment]
```
$ ./deviceQuery
./deviceQuery Starting...

 CUDA Device Query (Runtime API) version (CUDART static linking)

Detected 1 CUDA Capable device(s)

Device 0: "GeForce GT 750M"
  CUDA Driver Version / Runtime Version          8.0 / 8.0
  CUDA Capability Major/Minor version number:    3.0
  Total amount of global memory:                 2048 MBytes (2147024896 bytes)
  ( 2) Multiprocessors, (192) CUDA Cores/MP:     384 CUDA Cores
  GPU Max Clock rate:                            926 MHz (0.93 GHz)
  Memory Clock rate:                             2508 Mhz
  Memory Bus Width:                              128-bit
  L2 Cache Size:                                 262144 bytes
  Maximum Texture Dimension Size (x,y,z)         1D=(65536), 2D=(65536, 65536), 3D=(4096, 4096, 4096)
  Maximum Layered 1D Texture Size, (num) layers  1D=(16384), 2048 layers
  Maximum Layered 2D Texture Size, (num) layers  2D=(16384, 16384), 2048 layers
  Total amount of constant memory:               65536 bytes
  Total amount of shared memory per block:       49152 bytes
  Total number of registers available per block: 65536
  Warp size:                                     32
  Maximum number of threads per multiprocessor:  2048
  Maximum number of threads per block:           1024
  Max dimension size of a thread block (x,y,z): (1024, 1024, 64)
  Max dimension size of a grid size    (x,y,z): (2147483647, 65535, 65535)
  Maximum memory pitch:                          2147483647 bytes
  Texture alignment:                             512 bytes
  Concurrent copy and kernel execution:          Yes with 1 copy engine(s)
  Run time limit on kernels:                     Yes
  Integrated GPU sharing Host Memory:            No
  Support host page-locked memory mapping:       Yes
  Alignment requirement for Surfaces:            Yes
  Device has ECC support:                        Disabled
  Device supports Unified Addressing (UVA):      Yes
  Device PCI Domain ID / Bus ID / location ID:   0 / 1 / 0
  Compute Mode:
     < Default (multiple host threads can use ::cudaSetDevice() with device simultaneously) >

deviceQuery, CUDA Driver = CUDART, CUDA Driver Version = 8.0, CUDA Runtime Version = 8.0, NumDevs = 1, Device0 = GeForce GT 750M
Result = PASS
```

**Comment by sgpyc** [comment]
I think somewhere in the Makefile has a line to enable compilation for 3.0 device, it need to be enabled. So we are looking for support of device {3.0 (latest Mac with CUDA compatible GPU), 3.5 (K40c), and 6.x (Pascal generation)} and CUDA version {7.5 (latest stable release), 8.0} on {Linux, OS X}?

**Comment by neoblizz** [comment]
@jowens please make sure `GUNROCK_GENCODE_SM30` https://github.com/gunrock/gunrock/blob/dev/CMakeLists.txt#L170 is on and every other GENCODE in CMakeLists is turned off. GT 750M is CUDA compatibility 3.0.

**Comment by jowens** [comment]
oh, wow, i really did not expect that compiling for 3.0 would make a difference. but it does!

```
$ ctest
Test project /Users/jowens/Documents/working/gunrock-build
      Start  1: TEST_BFS
 1/11 Test  #1: TEST_BFS .........................   Passed    0.27 sec
      Start  2: TEST_BC
 2/11 Test  #2: TEST_BC ..........................   Passed    0.21 sec
      Start  3: TEST_CC
 3/11 Test  #3: TEST_CC ..........................   Passed    0.25 sec
      Start  4: TEST_SSSP
 4/11 Test  #4: TEST_SSSP ........................   Passed    0.21 sec
      Start  5: TEST_PAGERANK
 5/11 Test  #5: TEST_PAGERANK ....................   Passed    0.45 sec
      Start  6: TEST_TOPK
 6/11 Test  #6: TEST_TOPK ........................   Passed    0.20 sec
      Start  7: SHARED_LIB_TEST_BFS
 7/11 Test  #7: SHARED_LIB_TEST_BFS ..............   Passed    0.23 sec
      Start  8: SHARED_LIB_TEST_BC
 8/11 Test  #8: SHARED_LIB_TEST_BC ...............   Passed    0.19 sec
      Start  9: SHARED_LIB_TEST_CC
 9/11 Test  #9: SHARED_LIB_TEST_CC ...............   Passed    0.19 sec
      Start 10: SHARED_LIB_TEST_SSSP
10/11 Test #10: SHARED_LIB_TEST_SSSP .............***Failed  Required regular expression not found.Regex=[Node_ID.*1.*: Label.*39.*
]  0.19 sec
      Start 11: SHARED_LIB_TEST_PAGERANK
11/11 Test #11: SHARED_LIB_TEST_PAGERANK .........   Passed    0.20 sec
```

**Comment by jowens** [comment]
Here's the failure:

```
10/11 Test #10: SHARED_LIB_TEST_SSSP .............***Failed  Required regular expression not found.Regex=[Node_ID.*1.*: Label.*39.*
]  0.20 sec
Round 1 of sssp.
0	 1	 0	 queue3 	 oversize :	 10 ->	 11
Node_ID [0] : Label [51]
Node_ID [1] : Label [0]
Node_ID [2] : Label [57]
Node_ID [3] : Label [67]
Node_ID [4] : Label [17]
Node_ID [5] : Label [67]
Node_ID [6] : Label [76]

      Start 11: SHARED_LIB_TEST_PAGERANK
```

**Comment by neoblizz** [comment]
Under your build directory, there should be a new folder called `Testing` and in it should be LastTest log file, can we see that for shared library sssp test? It should have more information on why that test failed.

**Comment by jowens** [comment]
[LastTest.log.txt](https://github.com/gunrock/gunrock/files/468104/LastTest.log.txt)

**Comment by jowens** [comment]
(but it's the same as you see that I posted above)

**Comment by jowens** [comment]
Does seem a little weird if it passes `TEST_SSSP` but not `SHARED_LIB_TEST_SSSP`.

**Comment by jowens** [comment]
... but it's a different test set, so perhaps it is not so weird.

**Comment by neoblizz** [comment]
It's not that weird. Just ran it on my machine (Ubuntu w/ 3.5 CUDA compatibility and CUDA 7.5) and it failed;
```
      Start 11: SHARED_LIB_TEST_SSSP
11/12 Test #11: SHARED_LIB_TEST_SSSP .............***Failed  Required regular expression not found.Regex=[Node_ID.*1.*: Label.*39.*
]  0.14 sec
```
So, something changed in the last commits because it used to pass just fine. I am looking into this.

**Comment by neoblizz** [comment]
@yzhwang I think this is due to the multi-pass configuration; https://github.com/gunrock/gunrock/blob/dev_debug/shared_lib_tests/shared_lib_sssp.c#L18 changing `srcs` to `NULL` should fix it.
```
    struct GRSetup *config = InitSetup(1, NULL);   // gunrock configurations
```

**Comment by neoblizz** [comment]
@jowens tested. Fixed.

```
Test project /gunrock/env
      Start  1: TEST_BFS
 1/12 Test  #1: TEST_BFS .........................   Passed    6.51 sec
      Start  2: TEST_BC
 2/12 Test  #2: TEST_BC ..........................   Passed    2.00 sec
      Start  3: TEST_CC
 3/12 Test  #3: TEST_CC ..........................   Passed    0.52 sec
      Start  4: TEST_SSSP
 4/12 Test  #4: TEST_SSSP ........................   Passed    2.44 sec
      Start  5: TEST_PAGERANK
 5/12 Test  #5: TEST_PAGERANK ....................   Passed    3.51 sec
      Start  6: TEST_TOPK
 6/12 Test  #6: TEST_TOPK ........................   Passed    0.79 sec
      Start  7: TEST_MST
 7/12 Test  #7: TEST_MST .........................   Passed    1.37 sec
      Start  8: SHARED_LIB_TEST_BFS
 8/12 Test  #8: SHARED_LIB_TEST_BFS ..............   Passed    5.83 sec
      Start  9: SHARED_LIB_TEST_BC
 9/12 Test  #9: SHARED_LIB_TEST_BC ...............   Passed    0.14 sec
      Start 10: SHARED_LIB_TEST_CC
10/12 Test #10: SHARED_LIB_TEST_CC ...............   Passed    0.14 sec
      Start 11: SHARED_LIB_TEST_SSSP
11/12 Test #11: SHARED_LIB_TEST_SSSP .............   Passed    0.14 sec
      Start 12: SHARED_LIB_TEST_PAGERANK
12/12 Test #12: SHARED_LIB_TEST_PAGERANK .........   Passed    0.14 sec

100% tests passed, 0 tests failed out of 12

Total Test time (real) =  23.54 sec
```

I can't push the changes from the station I am working on.

**Comment by jowens** [comment]
@yzhwang please chime in when you have a chance!

**Comment by jowens** [comment]
@neoblizz maybe submit a pull request for @yzhwang so that he can look at it and approve it at the same time.

**Comment by yzhwang** [comment]
Thanks @neoblizz . You are right that the problem was due to multi-pass SSSP. But actually you do not need to disable it since the multi-pass works fine. The reason it failed is because we only used simple regex which assume that the source node is 0 (and when you set the source array to NULL, it will be set to 0 by default) and when I was testing multi-pass feature, I accidentally put 1 there, so it used node 1 as the source node and gave the incorrect answer. The simple fix would be to set the source array to {0}. I will do it on dev_debug and cherry pick it to pre-release. But thanks for quickly spotting where the problem is!
